lib/rucio/common/utils.py

```diff
@@ -611,16 +611,13 @@ def default(self, obj):  # pylint: disable=E0202
         return json.JSONEncoder.default(self, obj)
 
 
-def render_json(**data):
-    """ JSON render function
+def render_json(list_=None, **data: Optional[Any]):
```

Comment:
This approach works, and I think it's clearly readable. 

The only issue I see that could arise is that `data` could have a key called `list_` which could clash with the `list_` positional argument, e.g.:
```
render_json([3,4,5], **{'a': 3, 'b': 5, 'list_': [9,8,10]})
```

which results in:

```
TypeError: render_json() got multiple values for argument 'list_'
```

In practice, this **should** never happen, as we either pass a list or pass a dict; however, if we want to be extra safe and avoid erroring out in this edge case, you could rewrite the method in this way:

```python
def render_json(*args, **kwargs) -> str:
    """ Render a list or a dict as a JSON-formatted string. """
    if args and isinstance(args[0], list):
        data = args[0]
    elif kwargs:
        data = kwargs
    else:
        raise ValueError("Error while serializing object to JSON-formatted string: supported input types are list or dict.")
    return json.dumps(data, cls=APIEncoder)
```

which supports all our current use cases and doesn't error out in that edge case:

```
print(render_json([3,4,5]))
print(render_json(**{'a': 3, 'b': 5}))
print(render_json([3,4,5], **{'a': 3, 'b': 5, 'list_': [9,8,10]}))

[3, 4, 5]
{"a": 3, "b": 5}
[3, 4, 5]
```

Reply:
I can perform this change, but are you sure this is the desired behavior? 
With this implementation, if I pass multiple lists or a list followed by anything, only the first list will be serialized, throwing away the rest of parameters.

```
print(render_json([3, 4, 5]))
print(render_json([3, 4, 5], [6, 7, 8]))
print(render_json([3,4,5], **{'a': 3, 'b': 5, 'list_': [9,8,10]}))

[3, 4, 5]
[3, 4, 5]
[3, 4, 5]
```

Maybe you meant:
```
def render_json(*args, **kwargs) -> str:
   ...
    if args and isinstance(args[0], list):
        data = args # <-- Look this line
    elif kwargs:
    ...
```
This solution however does not solve the problem of a list followed by a dict parameter.

Maybe we can do something like: 
```
def render_json(*list_, **dict_) -> str:
    """ Render a list or a dict as a JSON-formatted string.
    """
    serialized_result = ""
    if list_ and isinstance(list_[0], list):
        serialized_result += json.dumps(list_, cls=APIEncoder)
    
    if dict_:
        serialized_result += json.dumps(dict_, cls=APIEncoder)
    
    if (not(list_ and isinstance(list_[0], list)) and not(dict_)):
        raise ValueError("Error while serializing object to JSON-formatted string: supported input types are list or dict.")
    
    return serialized_result
```

This last case may be a little less elegant, but handles all these cases:

```
print(render_json([3, 4, 5]))
print(render_json([3, 4, 5], [6, 7, 8]))
print(render_json(**{'a': 3, 'b': 5}))
print(render_json(**{'a': 3, 'b': 5, 'list_': [9,8,10]}))
print(render_json([3,4,5], **{'a': 3, 'b': 5, 'list_': [9,8,10]}))

[[3, 4, 5]]
[[3, 4, 5], [6, 7, 8]]
{"a": 3, "b": 5}
{"a": 3, "b": 5, "list_": [9, 8, 10]}
[[3, 4, 5]]{"a": 3, "b": 5, "list_": [9, 8, 10]}
```

Reply:
> With this implementation, if I pass multiple lists or a list followed by anything, only the first list will be serialized, throwing away the rest of parameters.

I would say that's the intended functionality, at least for the way that we're currently using these `render_json*` methods: we either render a single list or a dict of kwargs. The change I proposed was just to avoid the `TypeError` in the edge case where someone might pass in both a list and kwargs, and one of the kwargs happens to be `list_` - this edge case goes against the intended use case and should never happen in the first place, but it's better to be safe and avoid having a runtime error here.

I get your point though, so I think it might be a good idea to have someone else give their opinion on this - I will ask someone in the team to chime in

Reply:
Sure, I was just giving my 2 cents, I am not the expert here.
Anyway, since this may take a while, would you mind assigning the related #6598 issue to me?

Reply:
> Sure, I was just giving my 2 cents, I am not the expert here. Anyway, since this may take a while, would you mind assigning the related #6598 issue to me?

It doesn't let me do that. According to https://stackoverflow.com/a/41172554, it seems like you should comment on that issue first, and it can then be assigned to you